refactor:Code Organization (#30)
fsprojects/fsharp-cheatsheet: docs/fsharp-cheatsheet.md

<a name="CodeOrganization_Modules"></a>
### Modules

> Modules are key building blocks for grouping related code; they can contain `types`, `let` bindings, or (nested) sub `module`s.

Reviewer: maybe namespaces needs to go first? Maybe you need to mention that everything lives in hierarchical namespaces with dots, and you can `open` namespaces, modules, types to bring stuff into scope to allow relative references?

Author: That was intentional; I put namespaces second so I had something to work with.

Reviewer: There's definitely a chicken and egg aspect
I guess I'm looking for something to convey that:
- .NET itself only has types with dotted prefixes
- `module`s and `type`s are just two flavors of types
But, namespace and module have comonalities in terms of how you can use them a) within a source file b) in an overall system to group/nest things
It's pretty good relative to were it is and I do know that I'd turn the cheatsheet into a book very fast if I followed by natural tendencies when writing!

Identifiers within modules can be referenced using dot notation, or you can bring them into scope via the [`open`](#CodeOrganization_OpenAndAutoOpen) keyword.
Illustrative-only example:

    module Money =
        type CardInfo =

>             { number: string

Reviewer: do you indent by 2 or 4?

Author: I don't really have a preference, I use the defaults, which I guess is 4 here.

              expiration: int * int }

        type Payment =
            | Card of CardInfo
            | Cash of int

        module Functions =
            let getCreditCardType (cardNumber: string) =

>                 match cardNumber.ToCharArray() |> Array.truncate 2 with

Reviewer: this is a good example to put somehere but its too long and a bit of a non sequiter here for explaining namespacing/opens. Also can trucate throw?

Author: Agreed. I'll shorten it up.

truncate: I'll test, but I chose it over take, which does throw.

Reviewer: I can never remember truncate vs take but you are correct. Examples is still long for exmplaining namespaces tho!

                | [| '3'; '4' |] | [| '3'; '7' |]      -> "Amex"
                | [| '4'; _ |]                         -> "Visa"
                | [| '5'; x |] when '0' < x && x < '6' -> "MasterCard"
                | [| '6'; x |] when x = '0' || x = '5' -> "Discover"
                | _                                    -> "Unknown"

If there is only one module in a file, the `module` name can be declared at the top, and all code constructs
within the file will be included in the `module`s definition (no indentation required).

    module Functions  // notice there is no '=' when at the top of a file

    let sumOfSquares n = seq {1..n} |> Seq.sumBy (fun x -> x * x)  // Functions.sumOfSquares

### Namespaces
Namespaces are simply dotted names that prefix `type` and `module` declarations to allow for hierarchical scoping.
The first `namespace` directives must be placed at the top of the file. Subsequent `namespace` directives either:

> (a) create a sub-namespace (keeping prior declarations in scope); or (b) create a new namespace (and all prior declarations are discarded from scope).

Reviewer: you sure there are subnamespaces (keeping decls in scope)?

Author: Dang it. You're right.

    namespace MyNamespace

    module MyModule =  // MyNamspace.MyModule
        let myLet = ...  // MyNamspace.MyModule.myLet

    namespace MyNamespace.SubNamespace

    namespace MyNewNamespace  // a new namespace

> A top-level [`module`](#CodeOrganization_Modules) declaration can bear its namespace as a dotted prefix:

Reviewer: ```suggestion
A top-level [`module`](#CodeOrganization_Modules)'s namespace can be specified via a dotted prefix:
```

    module MyNamespace.SubNamespace.Functions

<a name="CodeOrganization_OpenAndAutoOpen"></a>
### Open and AutoOpen

The `open` keyword can be used on `module`, `namespace`, and `type`.

    module Groceries =
        type Fruit =
            | Apple
            | Banana

    let fruit1 = Groceries.Banana
    open Groceries  // module
    let fruit2 = Apple
    ---
    open System.Diagnostics  // namespace
    let stopwatch = Stopwatch.StartNew()  // Stopwatch is accessible
    ---
    open type System.Text.RegularExpressions.Regex  // type
    let isHttp url = IsMatch("^https?:", url)  // Regex.IsMatch directly accessible

> Available to `module` declarations only, is the `AutoOpen` attribute, which alleviates the need for an `open`.

Reviewer: ```suggestion
`module`s can be tagged with an `AutoOpen` Attribute, which immediately brings the contents into the containing scope. Care is advised though - the `module` will also be implicitly `open`ed when its parent is.
```

Not sure how to word it, but it would be great to convey its raison d'etre here, which for me is more or less as a way to do `let`s in namespaces.

    [<AutoOpen>]
    module Groceries =
        type Fruit =
            | Apple
            | Banana

    let fruit = Banana

*However*, `AutoOpen` should be used cautiously. When an `open` or `AutoOpen` is used, all declarations in the containing element
will be brought into scope. This can lead to [shadowing](https://en.wikipedia.org/wiki/Variable_shadowing); where the last

> named declaration replaces all prior identically-named declarations. There is *no* error - or even a warning - in F#, when shadowing occurs.

Reviewer: This secontion is pretty good now.... BUT!

maybe shadowing deserves an individual section? you used `open type` above for another reason, but for me a key thing that trips people up is how DU cases shadow others and/or type names. I always used to open a DU's module, but really `open type` on the DU brings less stuff in.

Not sure the "AuotOpen would sidestep thi"s is going to be grokked by someon scanning a cheatsheet...

Author: Maybe AutoOpen doesn't need a place on a cheat sheet?

Reviewer: That may well be the answer (not mentioning it, and conserving space for better things)

However for me AutoOpen can be an important part of the overall tools in the toolbox, e.g.:
- if you want to split something over 2 files (or 2 projects), but still have them in the same namespace, you can't have two "partial" modules with the same name
- so you use a namespace
- but then you can't have helper functions at top level
- so AutoOpen allows each file to contribute helpers to the top level of the namespace

I'm really not sure what the solution is, and I can't think of a sufficiently succinct example. I guess the main thing is not to give it prominence without conveying that it's way down the list in terms of being the answer to problems...

Author: Thanks

A [coding convention](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/conventions#sort-open-statements-topologically) exists for `open`
statements to avoid pitfalls; `AutoOpen` would sidestep this.

### Accessibility Modifiers

> F# supports `public`, `private` (limiting access to its containing `type` or `module`) and `internal` (limiting access to its containing assembly).

Reviewer: this is pretty tight now, nice

|-----------------------------------------------------------------|--------------------------------------------|
| Module                                                          | `module internal MyModule =`               |
| Module .. `let`                                                 | `let private value =`                      |
| Record                                                          | `type internal MyRecord = { id: int }`     |
| Record [ctor](#CodeOrganization_SmartConstructors)              | `type MyRecord = private { id: int }`      |
| Discriminated Union                                             | `type internal MyDiscUni = A \| B`         |
| Discriminated Union [ctor](#CodeOrganization_SmartConstructors) | `type MyDiscUni = private A \| B `         |
| Class                                                           | `type internal MyClass() =`                |
| Class [ctor](#CodeOrganization_SmartConstructors)               | `type MyClass private () =`                |
| Class Additional [ctor](#CodeOrganization_SmartConstructors)    | `internal new() = MyClass("defaultValue")` |

> | Class .. `let`                                                  | *Always private. Cannot be modified*       |

Reviewer: ```suggestion
| Class .. `let`                                                  | *Always private. Cannot be overridden*       |
```
(I would want to slip in a reference to `member val`, but maybe best to leave it!)

Author: Yeah, wasn't sure how far to take it. `val`s feel dirty.

> | `type` .. `member`                                              | `member private _.typeMember =`            |

Reviewer: ```suggestion
| `type` .. `member`                                              | `member private _.TypeMember =`            |
```

| `type` .. `val`                                                 | `val internal explicitInt : int`           |

> <a name="CodeOrganization_SmartConstructors"></a>

Reviewer: Nicely done!

##### Smart Constructors

Making a primary constructor (ctor) `private` or `internal` is a common convention for ensuring value integrity;
otherwise known as ["making illegal states unrepresentable" (YouTube:Effective ML)](https://youtu.be/-J8YyfrSwTk?si=ml3AWro6jG77F0YW&t=1080).

Example of Single-case Discriminated Union with a `private` constructor that constrains a quantity between 0 and 100:

    type UnitQuantity =
        private UnitQuantity of int
    
    module UnitQuantity =  // common idiom: type companion module
        let tryCreate qty =
            if qty < 1 || qty > 100
            then None
            else Some (UnitQuantity qty)
        let value (UnitQuantity uQty) = uQty
        let zero = UnitQuantity 0
    ...
    let unitQtyOpt = UnitQuantity.tryCreate 5

    let validQty =
        unitQtyOpt
        |> Option.defaultValue UnitQuantity.zero

### Recursive Reference

> F#'s type inference and name resolution runs in file and line order. By default, any forward references are considered errors.

Reviewer: also left to right, e.g. the `(int, values) ||> X.fold (fun ...)` type inference thing mght be worth a mention somewhere
obv this is nice and focusd so not here
